### DISCUSSION TITLE
Custom aggregate function

### DISCUSSION THREAD

**Discussion by biggerlambda** [duckdb/duckdb#3605] [General]
I want to write a custom aggregate function that I can use over a window. Is there an example of how I can add a custom aggregate that can be called as a window function? Do I have to write that in c++ ? 

Thanks in advance for any pointers .

**Comment by pdet** [comment]
Hi @biggerlambda, 

one option is to write your own extension. Extensions are useful for expanding DuckDB's functions and scanners to solve domain-specific issues. I'm currently working on a blogpost that depicts how to do that. In the meantime you can check this repo:

https://github.com/pdet/Scrooge-McDuck

Particularly the First and Last functions are aggregations functions that can be used on window functions so that might be a close match to your own use case.

**Comment by hawkfish** [comment]
Hi @biggerlambda - I wrote up some [documentation](https://github.com/duckdb/duckdb/tree/master/src/core_functions/aggregate) on this the other day. Basically, unless you are doing something with huge states (e.g., holistic aggregates like `median`), you can just define the `combine` method and we will construct segment trees for very efficient moving aggregations. If not, you can always defined a `window` method, but I would avoid that unless it is really necessary.
